include/envoy/thread/thread.h

```diff
@@ -2,40 +2,120 @@
 
 #include "envoy/common/pure.h"
 
+#include "absl/base/thread_annotations.h"
+
 namespace Envoy {
 namespace Thread {
 
 /**
- * Like the C++11 "basic lockable concept" but a concrete interface vs. a template.
+ * Like the C++11 "basic lockable concept" but a pure virtual interface vs. a template, and
+ * with thread annotations.
  */
-class BasicLockable {
+class LOCKABLE BasicLockable {
 public:
   virtual ~BasicLockable() {}
 
-  virtual void lock() PURE;
-  virtual bool try_lock() PURE;
-  virtual void unlock() PURE;
+  virtual void lock() EXCLUSIVE_LOCK_FUNCTION() PURE;
+  virtual bool tryLock() EXCLUSIVE_TRYLOCK_FUNCTION(true) PURE;
+  virtual void unlock() UNLOCK_FUNCTION() PURE;
 };
 
 /**
  * A lock guard that deals with an optional lock.
  */
-template <class T> class OptionalLockGuard {
+class SCOPED_LOCKABLE OptionalLockGuard {
 public:
-  OptionalLockGuard(T* lock) : lock_(lock) {
-    if (lock) {
-      lock->lock();
+  /**
+   * Establishes a scoped mutex-lock. If non-null, the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  OptionalLockGuard(BasicLockable* lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(lock) {
+    if (lock_) {
+      lock_->lock();
     }
   }
 
-  ~OptionalLockGuard() {
+  /**
+   * Destruction of the OptionalLockGuard unlocks the lock, if it is non-null.
+   */
+  ~OptionalLockGuard() UNLOCK_FUNCTION() {
     if (lock_) {
       lock_->unlock();
     }
   }
 
 private:
-  T* lock_;
+  BasicLockable* lock_;
+};
+
+/**
+ * A lock guard that deals with a lock that is not locked on construction, although
+ * it is unlocked on destruction, if necessary.
+ */
+class DeferredLockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock; the mutex is not locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  DeferredLockGuard(BasicLockable& lock) : lock_(lock), is_locked_(false) {}
+
+  /**
+   * Destruction of the DeferredLockGuard unlocks the lock, if it was locked.
+   */
+  ~DeferredLockGuard() {
+    if (is_locked_) {
+      lock_.unlock();
+    }
+  }
+
+  // Attempts to lock the mutex, if present. Returns false if no lock was taken.
+  bool tryLock() {
+    is_locked_ = lock_.tryLock();
+    return is_locked_;
+  }
+
+private:
+  BasicLockable& lock_;
+  bool is_locked_;
+};
+
+/**
+ * Implements a LockGuard that is identical to absl::ReleasableMutexLock, but takes a
+ * BasicLockable& to allow usages to be agnostic to cross-process mutexes vs. single-process
+ * mutexes.
+ */
+class SCOPED_LOCKABLE LockGuard {
+public:
+  /**
+   * Establishes a scoped mutex-lock; the mutex is locked upon construction.
+   *
+   * @param lock the mutex.
+   */
+  explicit LockGuard(BasicLockable& lock) EXCLUSIVE_LOCK_FUNCTION(lock) : lock_(&lock) {
+    lock_->lock();
+  }
+
+  /**
+   * Destruction of the LockGuard unlocks the lock, if it has not already been explicitly unlocked.
+   */
+  ~LockGuard() UNLOCK_FUNCTION() { unlock(); }
```

Comment:
I don't think this should be annotated as UNLOCK_FUNCTION.  From https://clang.llvm.org/docs/ThreadSafetyAnalysis.html:

> Previously: EXCLUSIVE_LOCK_FUNCTION, SHARED_LOCK_FUNCTION, UNLOCK_FUNCTION
RELEASE and RELEASE_SHARED declare that the function releases the given capability. The caller must hold the capability on entry, and will no longer hold it on exit. It does not matter whether the given capability is shared or exclusive.

This function may not hold the resource when it is called.